### PR TITLE
fix(connector): [Bluesnap] Throw proper error message for redirection scenario

### PR DESCRIPTION
**Pull request by srujanchikke**
## Type of Change

- [x] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This PR has following changes 
1)`CardHolderInfo` field is added in payment request of Bluesnap
2)Update error message and error code in statusUpdate in CallConnectorAction for redirection response

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

Previously for connector bluesnap we were not sending proper error message when 3ds is disabled in dashboard . This pr modifies statusUpdate in CallConnectorAction by making it as struct which contains error message and error code as well . This PR also have `cardHolderInfo` field added in payment request for bluesnap .

## How did you test it?

i tested it locally and ran unit tests for bluesnap

Screenshot 2023-06-06 at 4 00 36 PM

Screenshot 2023-06-06 at 8 15 38 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/checkout.rs [hunk 1]
```diff
@@ -1199,9 +1199,13 @@ impl services::ConnectorRedirectResponse for Checkout {
                 .change_context(errors::ConnectorError::ResponseDeserializationFailed)?;
         let connector_action = query
             .status
-            .map(|checkout_status| {
-                payments::CallConnectorAction::StatusUpdate(checkout_status.into())
-            })
+            .map(
+                |checkout_status| payments::CallConnectorAction::StatusUpdate {
+                    status: checkout_status.into(),
```
Comment (jagan-jaya) [inline crates/router/src/connector/checkout.rs:1204]: use from here
```suggestion
                    status: <Status>::from(checkout_status),
```

File: crates/router/src/connector/utils.rs [hunk 2]
```diff
@@ -453,17 +457,46 @@ impl CardData for api::Card {
...
+        let card_holder_name = self.card_holder_name.peek();
+        let words: Vec<&str> = card_holder_name.split_whitespace().collect();
+        if !words.is_empty() {
+            let first_name = Secret::new(if words.len() > 1 {
+                words[..words.len() - 1].join(" ")
+            } else {
+                words.first().unwrap_or(&"").to_string()
+            });
+            let last_name = Secret::new(if words.len() > 1 {
+                words.last().unwrap_or(&"").to_string()
```
Comment (jagan-jaya) [inline crates/router/src/connector/utils.rs:481]:
```suggestion
                words.last().unwrap_or_default()
```

File: crates/router/src/connector/utils.rs [hunk 3]
```diff
@@ -453,17 +457,46 @@ impl CardData for api::Card {
...
+    fn get_first_name_and_last_name_from_cardholdername(
+        &self,
+    ) -> Result<(Secret<String>, Secret<String>), errors::ConnectorError> {
+        let card_holder_name = self.card_holder_name.peek();
+        let words: Vec<&str> = card_holder_name.split_whitespace().collect();
+        if !words.is_empty() {
+            let first_name = Secret::new(if words.len() > 1 {
+                words[..words.len() - 1].join(" ")
+            } else {
+                words.first().unwrap_or(&"").to_string()
```
Comment (jagan-jaya) [inline crates/router/src/connector/utils.rs:478]:
```suggestion
                words.first().unwrap_or_default()
```

File: crates/router/src/services/api.rs [hunk 4]
```diff
@@ -190,8 +191,19 @@ where
...
+        payments::CallConnectorAction::StatusUpdate {
+            status,
+            code,
+            message,
+        } => {
             router_data.status = status;
+            let response = ErrorResponse {
+                code: code.unwrap_or_else(|| consts::NO_ERROR_MESSAGE.to_string()),
+                message: message.unwrap_or_else(|| consts::NO_ERROR_MESSAGE.to_string()),
+                status_code: 200,
```
Comment (jagan-jaya) [inline crates/router/src/services/api.rs:203]: Add a comment justifying why 200 is needed here

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/utils.rs (hunk 2 above)
Comment (srujanchikke) [reply crates/router/src/connector/utils.rs]: we can't use `unwrap_or_default()` here as the trait bound `&&str: std::default::Default` is not satisfied

File: crates/router/src/connector/utils.rs (hunk 3 above)
Comment (srujanchikke) [reply crates/router/src/connector/utils.rs]: same here

File: crates/router/src/connector/utils.rs [hunk 5]
```diff
@@ -453,17 +457,46 @@ impl CardData for api::Card {
...
             delimiter,
             self.card_exp_month.peek().clone()
         ))
     }
+
+    fn get_first_name_and_last_name_from_cardholdername(
+        &self,
+    ) -> Result<(Secret<String>, Secret<String>), errors::ConnectorError> {
+        let card_holder_name = self.card_holder_name.peek();
+        let words: Vec<&str> = card_holder_name.split_whitespace().collect();
```
[the remark this answers is not in the document]
Comment (srujanchikke) [reply crates/router/src/connector/utils.rs:473]: it works for " " too

**Review by jagan-jaya — DISMISSED** [review]

**Review by ArjunKarthik — DISMISSED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by jagan-jaya — APPROVED** [review]

**Review by vspecky — APPROVED** [review]

**Review by NarsGNA — APPROVED** [review]

File: crates/router/src/connector/bluesnap.rs [hunk 6]
```diff
@@ -1094,9 +1094,15 @@ impl services::ConnectorRedirectResponse for Bluesnap {
...
-            _ => Ok(payments::CallConnectorAction::StatusUpdate(
-                enums::AttemptStatus::AuthenticationFailed,
-            )),
+            _ => Ok(payments::CallConnectorAction::StatusUpdate {
+                status: enums::AttemptStatus::AuthenticationFailed,
+                error_code: redirection_result.code,
+                error_message: redirection_result
+                    .info
+                    .as_ref()
+                    .and_then(|info| info.errors.as_ref().and_then(|error| error.first()))
```
Comment (lsampras) [inline crates/router/src/connector/bluesnap.rs:1103]: what does first() here mean

I assume the structure is below one
```
    StatusUpdate {
        status: storage_enums::AttemptStatus,
        error_code: Option<String>,
        error_message: Option<String>,
    },
    ```

**Review by lsampras — APPROVED** [review]

**Pull request merged**